# Statistics Test 2.0

Time - 1.30 Hr


Total Marks = 60



**MCQ**

**Which of the following is a measure of central tendency?**

    A) Variance
    B) Standard Deviation
    C) Mean
    D) Range

Ans-> C) Mean

**In a dataset with values [2, 4, 4, 6, 8], what is the median?**

    A) 4
    B) 4.5
    C) 6
    D) 8

Ans-> A) 4

**Inferential statistics are used to:**

    A) Describe the data's central tendency
    B) Summarize the data
    C) Make predictions or inferences about a population based on a sample
    D) Visualize the data

Ans-> C) Make predictions or inferences about a population based on a sample

**What does a confidence interval represent?**

    A) The range within which the population mean is expected to fall with a certain probability
    B) The exact value of the population mean
    C) The variability within the sample
    D) The number of samples required

Ans-> A) The range within which the population mean is expected to fall with a certain probability

**According to the Central Limit Theorem, as the sample size increases, the distribution of the sample mean:**

    A) Becomes skewed
    B) Approaches a normal distribution
    C) Remains uniform
    D) Becomes bimodal

Ans-> B) Approaches a normal distribution

**The Central Limit Theorem is important because:**

    A) It helps in defining the exact population distribution
    B) It allows for the use of normal distribution to make inferences about sample means
    C) It eliminates the need for sampling
    D) It ensures sample sizes are always large

Ans-> B) It allows for the use of normal distribution to make inferences about sample means

**ANOVA is used to test:**

    A) The association between two categorical variables
    B) The differences between means of three or more groups
    C) The proportion of a sample
    D) The correlation between two continuous variables

Ans-> B) The differences between means of three or more groups

**A p-value less than 0.05 in an ANOVA test indicates:**

    A) There is no significant difference between group means
    B) The means of at least one group are significantly different from others
    C) The variances within groups are equal
    D) The sample size is too small

Ans-> B) The means of at least one group are significantly different from others

**The chi-square test of independence assesses:**

    A) If two continuous variables are related
    B) The goodness-of-fit of a model
    C) The association between two categorical variables
    D) The variance of a single variable

Ans-> C) The association between two categorical variables

**The two-sample t-test assumes:**

    A) Equal variances between the two groups
    B) Unequal variances between the two groups
    C) No assumption about variances
    D) The samples are paired

Ans-> A) Equal variances between the two groups

# Python Basic Exercise

### Solve any three

Que 1 -> You have a list a=[1,2,3,6,9,7,4,5,8], and you have to find the kth element from the list. This means I want to find the 2nd largest number from the list. Similarly, you have to create a function where I pass a number, say 4, and the function will return the 4th largest number from the list.

In [1]:
def find_kth_largest(a, k):
    a_sorted = sorted(a, reverse=True)
    return a_sorted[k-1]

a = [1, 2, 3, 6, 9, 7, 4, 5, 8]

k = 2
print(f"The {k}-th largest number is:", find_kth_largest(a, k))

k = 4
print(f"The {k}-th largest number is:", find_kth_largest(a, k))

The 2-th largest number is: 8
The 4-th largest number is: 6


Que 2-> you have a list [1,8,0,6,4,0,6,0,0,2,0], You need to shift all the zeros to the right , without using any inbuilt function.

In [2]:
def shift_zeros_to_right(lst):
    non_zero_elements = []
    zero_count = 0
    #creating new list
    # appending all non zero elements to it
    # counting the number of zeros parallely
    #then adding the zeros at end on new list
    for num in lst:
        if num != 0:
            non_zero_elements.append(num)
        else:
            zero_count += 1


    non_zero_elements.extend([0] * zero_count)
    return non_zero_elements


lst = [1, 8, 0, 6, 4, 0, 6, 0, 0, 2, 0]
result = shift_zeros_to_right(lst)
print("List after shifting zeros to the right:", result)

List after shifting zeros to the right: [1, 8, 6, 4, 6, 2, 0, 0, 0, 0, 0]


Que 3->Reverse a string without using any inbuilt function like slicing.

In [4]:
def reverse_string(s):
    char_list = list(s)
    left = 0
    right = len(char_list) - 1

    #swap the left and right
    while left < right:
        char_list[left], char_list[right] = char_list[right], char_list[left]
        left += 1
        right -= 1

    return ''.join(char_list)


s = "hello"
reversed_s = reverse_string(s)
print("Reversed string:", reversed_s)

Reversed string: olleh


Que 4-> To find common prefix in array of strings

array= ['Abce','Abcjhgj','Abcf']


The common prefix among the strings 'Abce', 'Abcjhgj', and 'Abcf' is 'Abc'. So, the common prefix is 'Abc'.

In [5]:
def find_common_prefix(arr):
    if not arr:
        return ""

    # Start with the first string as the initial common prefix
    # Iterate over the rest of the strings
    # Reduce the prefix length while it does not match the start of the current string
    prefix = arr[0]

    for string in arr[1:]:
        while not string.startswith(prefix):
            prefix = prefix[:-1]
            if not prefix:
                return ""

    return prefix


array = ['Abce', 'Abcjhgj', 'Abcf']
common_prefix = find_common_prefix(array)
print("Common prefix:", common_prefix)


Common prefix: Abc


Que 5- Given the list l = [1,2,0,1,2,0,0], I want the output to be ans = [0,0,0,2,2,1,1] without using any inbuilt function

In [9]:
def custom_sort(arr):

    count_0 = 0
    count_1 = 0
    count_2 = 0

    # Count occurrences of 0, 1, and 2
    # Rearrange the sorted list based on counts
    for num in arr:
        if num == 0:
            count_0 += 1
        elif num == 1:
            count_1 += 1
        elif num == 2:
            count_2 += 1


    sorted_arr = [0] * count_0 + [2] * count_2 + [1] * count_1

    return sorted_arr


l = [1, 2, 0, 1, 2, 0, 0]
sorted_list = custom_sort(l)
print("Sorted list:", sorted_list)

Sorted list: [0, 0, 0, 2, 2, 1, 1]


# Statistics Theory Question


Que 1- What is Central Limit Theoram ? Explain.

For a large enough sample size, the distribution of the sample mean will be approximately normal, even if the population distribution is not. The sample mean's distribution will have the same mean as the population mean and a standard deviation of the population standard deviation divided by the square root of the sample size.

Que 2- What is CDF(Comulative density function). Explain with mathematics.

The Cumulative Distribution Function (CDF) gives the probability that a random variable X is less than or equal to a certain value x.
- CDF(X) = P(X<=x) is probability that X will take value less than or equal to x

Que 3- What is Anova test. how we can calculate behind the scene ?

Annova tests if there are significant differences between the means of three or more groups.
- Steps
  - calculate group means
  - measure how group means differ from the overall mean (SSB)
  - measure variability within each group (SSW)
  - find 'Between grp mean Sq'(MSB=SSB/k-1) and 'Within grp mean Sq'(MSW=SSW/n-k)
  - F = (MSB/MSW)

Que 4-  What is Chi-Square test. how we will calculate behind the scene.

The Chi-Square test assesses how well observed data fits a theoretical distribution or tests the independence of two categorical variables.
- X^2 = summation [ (Oi-Ei)^2 /Ei ]
- to see the critical values using chi-square distribution table, we need 2 values
  - degree of freedom
  - alpha level
- if critical score> test score --> we cannot reject null hypothesis

Que 5-  What is the difference between inferential and descriptive statistics.

Descriptive Statistics:

  - Purpose: Summarize and describe the features of a data set.
  - Methods: Mean, median, mode, variance, standard deviation.
  - Application: Applied directly to the data you have.

Inferential Statistics:

  - Purpose: Make predictions or inferences about a population based on a sample.
  - Methods: Hypothesis testing
  - Application: Uses sample data to estimate and generalize to a larger population.

Que 6- Explain any 2 distribution.

1. Normal Distribution:

  - Shape: Symmetrical, bell-shaped curve.
  - Characteristics: Mean = Median = Mode; defined by its mean and standard deviation. It describes how data is distributed around the mean, with most values falling close to the mean and fewer values as you move away.

2. Uniform Distribution:

  - Shape: Flat, rectangular shape.
  - Characteristics: All outcomes are equally likely within a specified range. For example, in a continuous uniform distribution between and b, the probability density function is constant within this range and zero outside it.

# Statistics Mathematics  Question



Que 1 - if I generate heights for 1000 people from a normal distribution, what would be the probability for a new person to have a height of 70 inches using (CDF) then z-score table? take height in inches.

#### Mean height (in inches) of the population
mean_height = 66

#### Standard deviation of heights (in inches) of the population
std_dev_height = 3


##### Generate data using this code
###### Generate heights for 1000 people from a normal distribution
heights = np.random.normal(mean_height, std_dev_height, 1000)))

In [11]:
import scipy.stats as st

mean_height = 66
std_dev_height = 3
height = 70

# Calculating the  Z-score
z_score = (height - mean_height) / std_dev_height

# Calculate CDF for the Z-score
probability = st.norm.cdf(z_score)
print("Probability: ",probability*100,"%")

Probability:  90.87887802741321 %


Que 2-  Suppose a fitness center manager anticipates the attendance for a particular yoga class throughout the week. Based on historical data and other factors, they estimate the expected attendance in numbers:

### Expected Attendance:

Monday: 40

Tuesday: 45

Wednesday: 38

Thursday: 50

Friday: 55

Saturday: 60

Sunday: 65

However, after the week concludes, the actual attendance is recorded:



### Observed (Actual) Attendance:



Monday: 38

Tuesday: 43

Wednesday: 40

Thursday: 48

Friday: 52

Saturday: 58

Sunday: 62

Ho: There is no significant difference between the expected and observed attendance for the yoga class throughout the week.

Ha: There is a significant difference between the expected and observed attendance for the yoga class throughout the week.

#### Use chi-square test *Statistical Tools*

In [13]:
import scipy.stats as st

expected = [40, 45, 38, 50, 55, 60, 65]
observed = [38, 43, 40, 48, 52, 58, 62]

sum_observed = sum(observed)  # Should be 399
sum_expected = sum(expected)  # Should be 373

expected_normalized = [e * (sum_observed / sum_expected) for e in expected]

# Chi-Square Statistic and p-value
chi2_statistic, p_value = st.chisquare(f_obs=observed, f_exp=expected_normalized)
print(f"Chi-Square Statistic: {chi2_statistic}")
print(f"P-value: {p_value}")

Chi-Square Statistic: 0.3467729187346415
P-value: 0.9992367516752689


here p-value is greater than 0.05 --> hence we cannot reject null hypothesis

Que3 -> Calculate anova test





    | Student_ID | Semester | Total_Marks |
    |------------|----------|-------------|
    | 1          | Fall     | 85          |
    | 2          | Fall     | 78          |
    | 3          | Fall     | 92          |
    | 4          | Spring   | 88          |
    | 5          | Spring   | 75          |
    | 6          | Spring   | 82          |
    | 7          | Summer   | 90          |
    | 8          | Summer   | 86          |
    | 9          | Summer   | 79          |    


Null Hypothesis (Ho): The average marks of students in each semester are approximately equal.

Alternative Hypothesis (Ha): The average marks of students in at least one semester are different.

In [14]:
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols

data = {
    'Student_ID': [1, 2, 3, 4, 5, 6, 7, 8, 9],
    'Semester': ['Fall', 'Fall', 'Fall', 'Spring', 'Spring', 'Spring', 'Summer', 'Summer', 'Summer'],
    'Total_Marks': [85, 78, 92, 88, 75, 82, 90, 86, 79]
}
df = pd.DataFrame(data)

# ANOVA
model = ols('Total_Marks ~ Semester', data=df).fit()
anova_table = sm.stats.anova_lm(model, typ=1)

print(anova_table)

           df      sum_sq    mean_sq        F   PR(>F)
Semester  2.0   22.222222  11.111111  0.27248  0.77043
Residual  6.0  244.666667  40.777778      NaN      NaN


since p-value (0.77043) is greater than 0.05, we fail to reject the null hypothesis